tests_and_analysis/performance_benchmarking/artifact_handling/copy_benchmark_artifacts.py

```diff
@@ -0,0 +1,228 @@
+        type=str,
+        default="https://anvil.softeng-support.ac.uk/jenkins/job/"
+                "PACE-neutrons/job/Euphonic/job/Performance Benchmarking"
+                "/api/json/",
+        help="The url of the jenkins job running the performance benchmarking "
+             "that contains the artifacts to copy"
+    )
+    # Range of builds you wish to recover artifacts from
+    parser.add_argument(
+        "-r", "--range", action="store", nargs=2, dest="range", type=int,
+        help="The range of Jenkins builds you wish to recover artifacts from."
+    )
+    return parser
+
+
+def coerce(value: int, lower_bound: int, upper_bound: int) -> int:
+    """
+    Coerce the value into between the lower bound and upper bound (inclusive).
+
+    Parameters
+    ----------
+    value : int
+        The value to coerce
+    lower_bound : int
+        The inclusive lower bound to coerce the value into
+    upper_bound : int
+        The inclusive upper bound to coerce the value into
+
+    Returns
+    -------
+    int
+        The value coerced into being between the lower and upper bound.
+    """
+    return max(lower_bound, min(value, upper_bound))
+
+
+def coerce_range(performance_benchmarking_response_json: Dict,
+                 range_parsed: Union[None, List[int]]) -> Tuple[int, int]:
+    """
+    Take details of the builds and find the first and last build numbers.
+    Coerce the range parsed by the user into between those first and last build
+    numbers.
+
+    Parameters
+    ----------
+    performance_benchmarking_response_json : Dict
+        A json response from Jenkins with build details including firstBuild
+        and lastBuild
+    range_parsed : Union[None, List[int]]
+        If no range was parsed this will be None, if a range has been parsed
+        this will be a list of two integers.
+
+    Returns
+    -------
+    Tuple[int, int]
+        The first value is the range start point and the second is the range
+        end point.
+    """
+    # The first and last build numbers
+    largest_possible_range = (
+        performance_benchmarking_response_json["firstBuild"]["number"] - 1,
+        performance_benchmarking_response_json["lastBuild"]["number"]
+    )
+    if range_parsed:
+        # Force both build numbers to be within the boundaries
+        lower_build_number = coerce(
+            range_parsed[0],
+            largest_possible_range[0], largest_possible_range[1]
+        )
+        upper_build_number = coerce(
+            range_parsed[1],
+            largest_possible_range[0], largest_possible_range[1]
+        )
+        return lower_build_number, upper_build_number
+    else:
+        # When there is no range parsed copy all the builds
+        return largest_possible_range
+
+
+def write_to_file(directory: str, filename: str,
+                  performance_benchmarks_json: str):
+    # Check access
+    test_file = os.path.join(directory, "test.txt")
+    try:
+        open(test_file, "w+")
```

Comment:
Rather than creating a specific test file it might be better just to try and write the data that we want and let it fail if we don't have permissions - hopefully that should give a helpful error message anyway